File: sdks/php/src/ZVault.php
<?php

declare(strict_types=1);

namespace ZVault;

/**
 * ZVault PHP SDK — fetch and cache secrets from ZVault Cloud.
 *
 * Zero external dependencies — uses cURL.
 *
 * @example
 * $vault = new \ZVault\Client(getenv('ZVAULT_TOKEN'));
 * $secrets = $vault->getAll('production');
 * $dbUrl = $vault->get('DATABASE_URL', 'production');
 */
class Client
{
    private const DEFAULT_BASE_URL = 'https://api.zvault.cloud';
    private const DEFAULT_CACHE_TTL = 300; // 5 minutes
    private const DEFAULT_TIMEOUT = 10;
    private const MAX_RETRIES = 2;

    private string $token;
    private string $orgId;
    private string $projectId;
    private string $baseUrl;
    private int $cacheTtl;

    /** @var array<string, array{data: array<string, string>, expires_at: float}> */
    private array $cache = [];

    public function __construct(
        string $token,
        ?string $orgId = null,
        ?string $projectId = null,
        ?string $baseUrl = null,
        ?int $cacheTtl = null,
    ) {
        $this->token = $token ?: (getenv('ZVAULT_TOKEN') ?: '');
        $this->orgId = $orgId ?: (getenv('ZVAULT_ORG_ID') ?: '');
        $this->projectId = $projectId ?: (getenv('ZVAULT_PROJECT_ID') ?: '');
        $this->baseUrl = rtrim($baseUrl ?: (getenv('ZVAULT_URL') ?: self::DEFAULT_BASE_URL), '/');
        $this->cacheTtl = $cacheTtl ?? self::DEFAULT_CACHE_TTL;

        if (empty($this->token) || empty($this->orgId) || empty($this->projectId)) {
            throw new ZVaultException('token, orgId, and projectId are required');
        }
    }

    /**
     * Fetch all secrets for the given environment.
     *
     * @return array<string, string>
     * @throws ZVaultException
     */
    public function getAll(string $env = 'production'): array
    {
        if (isset($this->cache[$env]) && $this->cache[$env]['expires_at'] > microtime(true)) {
            return $this->cache[$env]['data'];
        }

        try {
            $url = sprintf(
                '%s/v1/cloud/orgs/%s/projects/%s/envs/%s/secrets',
                $this->baseUrl, $this->orgId, $this->projectId, $env
            );
            $body = $this->fetchWithRetry($url);
            $secrets = $this->parseSecrets($body);
            $this->cache[$env] = [
                'data' => $secrets,
                'expires_at' => microtime(true) + $this->cacheTtl,
            ];
            return $secrets;
        } catch (\Throwable $e) {
            if (isset($this->cache[$env])) {
                return $this->cache[$env]['data'];
            }
            throw new ZVaultException('Failed to fetch secrets: ' . $e->getMessage(), 0, $e);
        }
    }

    /**
     * Fetch a single secret by key.
     *
     * @throws ZVaultException
     */
    public function get(string $key, string $env = 'production'): string
    {
        $all = $this->getAll($env);
        if (!array_key_exists($key, $all)) {
            throw new ZVaultException("Secret not found: {$key}");
        }
        return $all[$key];
    }

    /**
     * Check if the ZVault API is reachable.
     */
    public function healthy(): bool
    {
        try {
            $ch = curl_init($this->baseUrl . '/health');
            if ($ch === false) return false;
            curl_setopt_array($ch, [
                CURLOPT_RETURNTRANSFER => true,
                CURLOPT_TIMEOUT => 5,
                CURLOPT_NOBODY => true,
            ]);
            curl_exec($ch);
            $code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
            curl_close($ch);
            return $code === 200;
        } catch (\Throwable) {
            return false;
        }
    }

    /**
     * Inject all secrets into $_ENV and putenv().
     *
     * @return int Number of secrets injected
     */
    public function injectIntoEnv(string $env = 'production'): int
    {
        $secrets = $this->getAll($env);
        $count = 0;
        foreach ($secrets as $k => $v) {
            if (getenv($k) === false) {
                putenv("{$k}={$v}");
                $_ENV[$k] = $v;
                $count++;
            }
        }
        return $count;
    }

    private function fetchWithRetry(string $url): string
    {
        $lastErr = null;

        for ($i = 0; $i <= self::MAX_RETRIES; $i++) {
            try {
                $ch = curl_init($url);
                if ($ch === false) {
                    throw new ZVaultException('Failed to init cURL');
                }

                curl_setopt_array($ch, [
                    CURLOPT_RETURNTRANSFER => true,
                    CURLOPT_TIMEOUT => self::DEFAULT_TIMEOUT,
                    CURLOPT_HTTPHEADER => [
                        'Authorization: Bearer ' . $this->token,
                        'Content-Type: application/json',
                        'User-Agent: zvault-php/0.1.0',
                    ],
                ]);

                $body = curl_exec($ch);
                $code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
                $err = curl_error($ch);
                curl_close($ch);

                if ($body === false) {
                    throw new ZVaultException("cURL error: {$err}");
                }

                if ($code >= 200 && $code < 300) {
                    return (string) $body;
                }

                $lastErr = new ZVaultException("HTTP {$code}");
                if ($code < 500 && $code !== 429) {
                    throw $lastErr;
                }
            } catch (ZVaultException $e) {
                throw $e;
            } catch (\Throwable $e) {
                $lastErr = $e;
            }

            if ($i < self::MAX_RETRIES) {
                usleep((int) (300_000 * (2 ** $i)));
            }
        }

        throw new ZVaultException('Request failed after retries: ' . ($lastErr?->getMessage() ?? 'unknown'));
    }

    /**
     * @return array<string, string>
     */
    private function parseSecrets(string $json): array
    {
        $data = json_decode($json, true);
        $result = [];
        foreach (($data['secrets'] ?? []) as $s) {
            if (isset($s['key'], $s['value'])) {
                $result[$s['key']] = $s['value'];
            }
        }
        return $result;
    }
}

class ZVaultException extends \RuntimeException {}

File: sdks/ruby/lib/zvault.rb
# frozen_string_literal: true

require "net/http"
require "json"
require "uri"

# ZVault Ruby SDK — fetch and cache secrets from ZVault Cloud.
#
# @example
#   vault = ZVault::Client.new(token: ENV["ZVAULT_TOKEN"])
#   secrets = vault.get_all(env: "production")
#   db_url = vault.get("DATABASE_URL", env: "production")
#
module ZVault
  class Error < StandardError; end

  class Client
    DEFAULT_BASE_URL = "https://api.zvault.cloud"
    DEFAULT_CACHE_TTL = 300 # 5 minutes
    DEFAULT_TIMEOUT = 10
    MAX_RETRIES = 2

    # @param token [String] Service token (or ZVAULT_TOKEN env var)
    # @param org_id [String] Organization ID (or ZVAULT_ORG_ID env var)
    # @param project_id [String] Project ID (or ZVAULT_PROJECT_ID env var)
    # @param base_url [String] API base URL
    # @param cache_ttl [Integer] Cache TTL in seconds
    def initialize(token: nil, org_id: nil, project_id: nil, base_url: nil, cache_ttl: nil)
      @token = token || ENV["ZVAULT_TOKEN"] || raise(Error, "token is required")
      @org_id = org_id || ENV["ZVAULT_ORG_ID"] || raise(Error, "org_id is required")
      @project_id = project_id || ENV["ZVAULT_PROJECT_ID"] || raise(Error, "project_id is required")
      @base_url = (base_url || ENV["ZVAULT_URL"] || DEFAULT_BASE_URL).chomp("/")
      @cache_ttl = cache_ttl || DEFAULT_CACHE_TTL
      @cache = {}
      @mutex = Mutex.new
    end

    # Fetch all secrets for the given environment.
    # @param env [String] Environment slug
    # @return [Hash<String, String>]
    def get_all(env: "production")
      @mutex.synchronize do
        entry = @cache[env]
        return entry[:data] if entry && entry[:expires_at] > Time.now

        begin
          url = "#{@base_url}/v1/cloud/orgs/#{@org_id}/projects/#{@project_id}/envs/#{env}/secrets"
          body = fetch_with_retry(url)
          secrets = parse_secrets(body)
          @cache[env] = { data: secrets, expires_at: Time.now + @cache_ttl }
          secrets
        rescue StandardError => e
          return entry[:data] if entry

          raise Error, "Failed to fetch secrets: #{e.message}"
        end
      end
    end

    # Fetch a single secret by key.
    # @param key [String] Secret key
    # @param env [String] Environment slug
    # @return [String]
    def get(key, env: "production")
      all = get_all(env: env)
      all.fetch(key) { raise Error, "Secret not found: #{key}" }
    end

    # Check if the ZVault API is reachable.
    # @return [Boolean]
    def healthy?
      uri = URI("#{@base_url}/health")
      res = Net::HTTP.get_response(uri)
      res.is_a?(Net::HTTPSuccess)
    rescue StandardError
      false
    end

    # Inject all secrets into ENV.
    # @param env [String] Environment slug
    # @return [Integer] Number of secrets injected
    def inject_into_env(env: "production")
      secrets = get_all(env: env)
      count = 0
      secrets.each do |k, v|
        unless ENV.key?(k)
          ENV[k] = v
          count += 1
        end
      end
      count
    end

    private

    def fetch_with_retry(url)
      last_err = nil

      (MAX_RETRIES + 1).times do |i|
        begin
          uri = URI(url)
          http = Net::HTTP.new(uri.host, uri.port)
          http.use_ssl = uri.scheme == "https"
          http.open_timeout = DEFAULT_TIMEOUT
          http.read_timeout = DEFAULT_TIMEOUT

          req = Net::HTTP::Get.new(uri)
          req["Authorization"] = "Bearer #{@token}"
          req["Content-Type"] = "application/json"
          req["User-Agent"] = "zvault-ruby/0.1.0"

          res = http.request(req)

          return res.body if res.is_a?(Net::HTTPSuccess)

          last_err = Error.new("HTTP #{res.code}")
          raise last_err if res.code.to_i < 500 && res.code.to_i != 429
        rescue Error
          raise
        rescue StandardError => e
          last_err = e
        end

        sleep(0.3 * (2**i)) if i < MAX_RETRIES
      end

      raise Error, "Request failed after retries: #{last_err&.message}"
    end

    def parse_secrets(json_str)
      data = JSON.parse(json_str)
      result = {}
      Array(data["secrets"]).each do |s|
        result[s["key"]] = s["value"] if s["key"] && s["value"]
      end
      result.freeze
    end
  end
end
